lib/backends/sdc/containers.js

```diff
@@ -559,60 +525,302 @@ function addRulesToFWAPI(opts, rules, payload, callback) {
+ *   With fabrics enabled:
+ *   2. 'bridge' or nothing specified provisions on the user's default network,
+ *      potentially also with the external network.
+ *   3. Specifying a network name will provision to the named *fabric* network.
+ *   4. Specifying a network id (or portion) will provision to that *fabric*
+ *      network.
+ *
+ *   Docker resolves name/id collisions in the following way:
+ *     - a name is preferred to a partial id
+ *     - a full id is preferred to a name
  */
 function addNetworksToPayload(opts, container, payload, callback) {
     assert.object(opts, 'opts');
     assert.object(opts.config, 'opts.config');
     assert.object(opts.config.napi, 'opts.config.napi');
     assert.object(opts.log, 'opts.log');
+
+    assert.object(opts.config.overlay);
+
+
+    assert.optionalString(opts.config.overlay.externalPool,
+        'opts.config.overlay.externalPool');
+    assert.optionalBool(opts.config.overlay.enabled,
+        'opts.config.overlay.enabled');
     assert.func(callback, 'callback');
 
-    if (opts.config.overlay.enabled) {
-        opts.log.debug('Fabrics configured: using for networking');
-        return addFabricNetworksToPayload(opts, container, payload, callback);
+    /*
+     * currently there are four network possibilities:
+     *  - no fabrics configured -> external network.
+     *  - fabrics configured and...
+     *    - 'bridge' or no network specified: default fabric and...
+     *      - if publishing ports or building: also external network
+     *    - fabric name specified: named fabric only
+     *    unsupported:
+     *
+     * We need to determine our network 'mode'. Docker defines the following
+     * networking modes:
+     *
+     * 1. 'legacy bridge' - the default currently supported.
+     * 2. 'host' - not meaningful for us
+     * 2. 'container' - like 'host', not meaningful
+     * 3. 'none' - eventually supportable
+     * 4. 'user-defined bridge' - probably supportable
+     * 5. 'user-defined overlay' - definitely supportable
+     *
+     */
+    if (!opts.config.overlay.enabled) {
+        return externalNetworkByName(opts, container, payload, callback);
+    } else {
+        if (container.HostConfig.NetworkMode === 'bridge'
+            || container.HostConfig.NetworkMode === 'default'
+            || common.objEmpty(container.HostConfig.NetworkMode)) {
+            return defaultFabricNetwork(opts, container, payload, callback);
+        } else {
+            return namedNetwork(opts, container, payload, callback);
+        }
     }
+}
+
+/*
+ * When fabrics are not configured, we will use the named
+ * network, defaulting to 'external'.
+ */
+function externalNetworkByName(opts, container, payload, callback) {
+    assert.object(opts.config, 'opts.config');
+    assert.optionalString(opts.config.externalNetwork,
+        'opts.config.externalNetwork');
+    assert.object(opts.account, 'opts.account');
+    assert.string(opts.account.uuid, 'opts.account.uuid');
 
```

Comment:
I added an assert on `payload`, since that's the only one we write to; `container` is unused, and so I removed it from the parameters instead.